src/override/malloc.cc

```diff
@@ -221,16 +297,267 @@ extern "C"
   // various events.  They are currently unused, though we may wish to reset
   // statistics on fork if built with statistics.
 
-  SNMALLOC_EXPORT void SNMALLOC_NAME_MANGLE(_malloc_prefork)(void) {}
-  SNMALLOC_EXPORT void SNMALLOC_NAME_MANGLE(_malloc_postfork)(void) {}
-  SNMALLOC_EXPORT void SNMALLOC_NAME_MANGLE(_malloc_first_thread)(void) {}
+  SNMALLOC_EXPORT SNMALLOC_USED_FUNCTION inline void _malloc_prefork(void) {}
+  SNMALLOC_EXPORT SNMALLOC_USED_FUNCTION inline void _malloc_postfork(void) {}
+  SNMALLOC_EXPORT SNMALLOC_USED_FUNCTION inline void _malloc_first_thread(void)
+  {}
+
+  /**
+   * Jemalloc API provides a way of avoiding name lookup when calling
+   * `mallctl`.  For now, always return an error.
+   */
+  int SNMALLOC_NAME_MANGLE(mallctlnametomib)(const char*, size_t*, size_t*)
+  {
+    return ENOENT;
+  }
 
+  /**
+   * Jemalloc API provides a generic entry point for various functions.  For
+   * now, this is always implemented to return an error.
+   */
+  int SNMALLOC_NAME_MANGLE(mallctlbymib)(
+    const size_t*, size_t, void*, size_t*, void*, size_t)
+  {
+    return ENOENT;
+  }
+
+  /**
+   * Jemalloc API provides a generic entry point for various functions.  For
+   * now, this is always implemented to return an error.
+   */
   SNMALLOC_EXPORT int
     SNMALLOC_NAME_MANGLE(mallctl)(const char*, void*, size_t*, void*, size_t)
   {
     return ENOENT;
   }
 
+  /**
+   * Jemalloc 3 experimental API.  Allocates  at least `size` bytes and returns
+   * the result in `*ptr`, if `rsize` is not null then writes the allocated size
+   * into `*rsize`.  `flags` controls whether the memory is zeroed and what
+   * alignment is requested.
+   */
+  int SNMALLOC_NAME_MANGLE(allocm)(
+    void** ptr, size_t* rsize, size_t size, int flags)
+  {
+    auto f = JEMallocFlags(flags);
+    size = f.aligned_size(size);
+    if (rsize != nullptr)
+    {
+      *rsize = size;
+    }
+    if (f.should_zero())
+    {
+      *ptr = ThreadAlloc::get().alloc<ZeroMem::YesZero>(size);
+    }
+    else
+    {
+      *ptr = ThreadAlloc::get().alloc(size);
+    }
+    return (*ptr != nullptr) ? allocm_success : allocm_err_oom;
+  }
+
+  /**
+   * Jemalloc 3 experimental API.  Reallocates the allocation in `*ptr` to be at
+   * least `size` bytes and returns the result in `*ptr`, if `rsize` is not null
+   * then writes the allocated size into `*rsize`.  `flags` controls whether the
+   * memory is zeroed and what alignment is requested and whether reallocation
+   * is permitted.  If reallocating, the size will be at least `size` + `extra`
```

Comment:
Not sure I understand the use of `extra` here?  Is the intended use, I need it to be `size` big, but if you need to reallocate, make it `extra` bigger.  I think your comment says that, but it took me a couple of passes.

Reply:
The actual requirement (from the half-sentence 'specification' in the jemalloc manual) for `extra` is that the caller would *like* `size + extra` bytes, but will accept anything above `size` bytes.  Since we're a sizeclass allocator, we always either return the current size if it's `>= size` or allocate to be at least `size+extra` here.